alacritty/src/macos/proc.rs

```diff
@@ -0,0 +1,119 @@
+#![allow(non_camel_case_types)]
+use std::ffi::{CStr, CString};
+use std::io;
+use std::mem::{self, MaybeUninit};
+use std::os::raw;
+use std::path::PathBuf;
+
+pub fn cwd(pid: raw::c_int) -> io::Result<PathBuf> {
+    let mut info = MaybeUninit::<proc_vnodepathinfo>::uninit();
+    let size = mem::size_of::<proc_vnodepathinfo>() as raw::c_int;
+    let c_str = unsafe {
+        match proc_pidinfo(
+            pid,
+            PROC_PIDVNODEPATHINFO,
+            0,
+            info.as_mut_ptr() as *mut raw::c_void,
+            size,
+        ) {
+            c if c < 0 => return Err(io::Error::last_os_error()),
+            s if s != size => {
+                return Err(io::Error::new(io::ErrorKind::Other, "invalid return size"))
+            },
+            _ => (),
+        }
+        CStr::from_ptr(info.assume_init().pvi_cdir.vip_path.as_ptr())
+    };
+
+    CString::from(c_str)
+        .into_string()
+        .map(PathBuf::from)
+        .map_err(|e| io::Error::new(io::ErrorKind::Other, e))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_cwd() {
+        let pid = std::process::id();
+        cwd(pid as i32)
+            .map(|path| assert!(!path.to_str().expect("UTF-8 encoding").is_empty()))
```

Comment:
Performing an assertion inline like this I would usually consider pretty bad practice. Generally `.map` should not be used as a `.for_each` kinda thing, but instead only be used when the goal is to transform one value into another.

So just doing `let cwd = cwd(pid).expect(...); assert!` should be much better.

It's also possible to just get the current working directory of the test process itself using `std::env::current_dir`, so I feel like we should be able to write a much better test with that instead? Then the whole expect stuff wouldn't even be necessary.

Reply:
Sadly `std::io::Error` doesn't implement `==`.
https://users.rust-lang.org/t/help-understanding-io-error-and-lack-of-partialeq/13212/9

Reply:
It should never error anyways, right? Might as well just map it or something. Though unwrapping is perfectly fine in tests too if it generates the simpler code. The errors for unwraps in test code are extremely easy to debug, so much so that I probably wouldn't even bother with `expect` unless there are multiple errors on the same line.